File: web-dashboard/src/AppLogger.js
// Sistema de logs centralizado para el dashboard web
const MAX_LOGS = 500
let logs = []
let listeners = []

export const LogLevel = {
  INFO: 'INFO',
  SUCCESS: 'SUCCESS',
  WARNING: 'WARNING',
  ERROR: 'ERROR',
  NETWORK: 'NETWORK',
  ALERT: 'ALERT',
  COMMAND: 'COMMAND',
  RESPONSE: 'RESPONSE'
}

const getIcon = (level) => {
  switch(level) {
    case LogLevel.INFO: return 'ℹ️'
    case LogLevel.SUCCESS: return '✅'
    case LogLevel.WARNING: return '⚠️'
    case LogLevel.ERROR: return '❌'
    case LogLevel.NETWORK: return '🌐'
    case LogLevel.ALERT: return '🚨'
    case LogLevel.COMMAND: return '📤'
    case LogLevel.RESPONSE: return '📥'
    default: return '•'
  }
}

const getColor = (level) => {
  switch(level) {
    case LogLevel.SUCCESS: return '#22c55e'
    case LogLevel.WARNING: return '#f59e0b'
    case LogLevel.ERROR: return '#ef4444'
    case LogLevel.ALERT: return '#ef4444'
    case LogLevel.NETWORK: return '#3b82f6'
    case LogLevel.COMMAND: return '#8b5cf6'
    case LogLevel.RESPONSE: return '#06b6d4'
    default: return '#ffffff'
  }
}

const formatTime = (timestamp) => {
  const date = new Date(timestamp)
  return date.toLocaleTimeString('es-AR', { 
    hour: '2-digit', 
    minute: '2-digit', 
    second: '2-digit',
    fractionalSecondDigits: 3
  })
}

const addLog = (level, tag, message, details = null) => {
  const entry = {
    id: Date.now() + Math.random(),
    timestamp: Date.now(),
    level,
    tag,
    message,
    details,
    icon: getIcon(level),
    color: getColor(level),
    formattedTime: formatTime(Date.now())
  }
  
  logs.push(entry)
  
  // Limitar tamaño
  while (logs.length > MAX_LOGS) {
    logs.shift()
  }
  
  // Notificar listeners
  listeners.forEach(listener => listener(entry))
  
  // También loguear en consola
  const consoleMethod = level === LogLevel.ERROR ? 'error' : 
                        level === LogLevel.WARNING ? 'warn' : 'log'
  console[consoleMethod](`[${tag}] ${message}`, details || '')
  
  return entry
}

// API pública
export const AppLogger = {
  getLogs: () => [...logs],
  
  clear: () => {
    logs = []
    listeners.forEach(listener => listener(null))
  },
  
  addListener: (listener) => {
    listeners.push(listener)
    return () => {
      listeners = listeners.filter(l => l !== listener)
    }
  },
  
  // Métodos de logging
  info: (tag, message, details) => addLog(LogLevel.INFO, tag, message, details),
  success: (tag, message, details) => addLog(LogLevel.SUCCESS, tag, message, details),
  warning: (tag, message, details) => addLog(LogLevel.WARNING, tag, message, details),
  error: (tag, message, details) => addLog(LogLevel.ERROR, tag, message, details),
  network: (tag, message, details) => addLog(LogLevel.NETWORK, tag, message, details),
  alert: (tag, message, details) => addLog(LogLevel.ALERT, tag, message, details),
  command: (tag, message, details) => addLog(LogLevel.COMMAND, tag, message, details),
  response: (tag, message, details) => addLog(LogLevel.RESPONSE, tag, message, details),
  
  // Métodos específicos para Supabase
  supabaseRequest: (endpoint, method, body) => {
    addLog(LogLevel.COMMAND, 'SUPABASE', `${method} ${endpoint}`, body)
  },
  
  supabaseResponse: (endpoint, status, response) => {
    if (status >= 200 && status < 300) {
      addLog(LogLevel.RESPONSE, 'SUPABASE', `✓ ${endpoint} -> ${status}`, response?.substring(0, 200))
    } else {
      addLog(LogLevel.ERROR, 'SUPABASE', `✗ ${endpoint} -> ${status}`, response)
    }
  },
  
  supabaseError: (endpoint, error) => {
    addLog(LogLevel.ERROR, 'SUPABASE', `✗ ${endpoint} FAILED`, error)
  },
  
  // Métodos específicos para alertas
  alertDetected: (deviceId, temp, message) => {
    addLog(LogLevel.ALERT, 'ALERT', `🚨 ALERTA DETECTADA: ${deviceId}`, `Temp: ${temp}°C - ${message}`)
  },
  
  alertSilenced: (deviceId, source) => {
    addLog(LogLevel.SUCCESS, 'ALERT', `🔕 ALERTA SILENCIADA: ${deviceId}`, `Origen: ${source}`)
  },
  
  silenceCommandSent: (deviceId) => {
    addLog(LogLevel.COMMAND, 'SILENCE', '📤 Enviando comando SILENCE', `Device: ${deviceId}`)
  },
  
  silenceCommandConfirmed: (deviceId) => {
    addLog(LogLevel.SUCCESS, 'SILENCE', '✓ Comando SILENCE confirmado', `Device: ${deviceId}`)
  },
  
  silenceCommandFailed: (deviceId, error) => {
    addLog(LogLevel.ERROR, 'SILENCE', '✗ Comando SILENCE falló', `Device: ${deviceId} - ${error}`)
  },
  
  // Polling
  pollingResult: (devicesCount, hasAlert) => {
    const status = hasAlert ? '⚠️ CON ALERTA' : '✅ Sin alertas'
    addLog(LogLevel.INFO, 'POLLING', `📊 ${devicesCount} dispositivo(s) - ${status}`)
  }
}

export default AppLogger

File: android-app-cloud/app/src/main/java/com/parametican/alertarift/AppLogger.kt
package com.parametican.alertarift

import android.util.Log
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Sistema de logs centralizado para debugging
 * Guarda todos los eventos importantes de la app
 */
object AppLogger {
    
    private const val MAX_LOGS = 500
    private val logs = CopyOnWriteArrayList<LogEntry>()
    private val dateFormat = SimpleDateFormat("HH:mm:ss.SSS", Locale.getDefault())
    
    data class LogEntry(
        val timestamp: Long,
        val level: LogLevel,
        val tag: String,
        val message: String,
        val details: String? = null
    ) {
        fun getFormattedTime(): String = dateFormat.format(Date(timestamp))
        
        fun getIcon(): String = when(level) {
            LogLevel.INFO -> "ℹ️"
            LogLevel.SUCCESS -> "✅"
            LogLevel.WARNING -> "⚠️"
            LogLevel.ERROR -> "❌"
            LogLevel.NETWORK -> "🌐"
            LogLevel.ALERT -> "🚨"
            LogLevel.COMMAND -> "📤"
            LogLevel.RESPONSE -> "📥"
        }
    }
    
    enum class LogLevel {
        INFO, SUCCESS, WARNING, ERROR, NETWORK, ALERT, COMMAND, RESPONSE
    }
    
    // Listeners para actualizar UI en tiempo real
    private val listeners = mutableListOf<(LogEntry) -> Unit>()
    
    fun addListener(listener: (LogEntry) -> Unit) {
        listeners.add(listener)
    }
    
    fun removeListener(listener: (LogEntry) -> Unit) {
        listeners.remove(listener)
    }
    
    private fun addLog(level: LogLevel, tag: String, message: String, details: String? = null) {
        val entry = LogEntry(System.currentTimeMillis(), level, tag, message, details)
        logs.add(entry)
        
        // Limitar tamaño
        while (logs.size > MAX_LOGS) {
            logs.removeAt(0)
        }
        
        // Notificar listeners
        listeners.forEach { it(entry) }
        
        // También loguear en Logcat
        when(level) {
            LogLevel.ERROR -> Log.e(tag, "$message ${details ?: ""}")
            LogLevel.WARNING -> Log.w(tag, "$message ${details ?: ""}")
            else -> Log.d(tag, "$message ${details ?: ""}")
        }
    }
    
    fun getLogs(): List<LogEntry> = logs.toList()
    
    fun getLogsFiltered(level: LogLevel? = null, tag: String? = null): List<LogEntry> {
        return logs.filter { entry ->
            (level == null || entry.level == level) &&
            (tag == null || entry.tag == tag)
        }
    }
    
    fun clear() {
        logs.clear()
    }
    
    // ============================================
    // MÉTODOS DE LOGGING POR CATEGORÍA
    // ============================================
    
    fun info(tag: String, message: String, details: String? = null) {
        addLog(LogLevel.INFO, tag, message, details)
    }
    
    fun success(tag: String, message: String, details: String? = null) {
        addLog(LogLevel.SUCCESS, tag, message, details)
    }
    
    fun warning(tag: String, message: String, details: String? = null) {
        addLog(LogLevel.WARNING, tag, message, details)
    }
    
    fun error(tag: String, message: String, details: String? = null) {
        addLog(LogLevel.ERROR, tag, message, details)
    }
    
    fun network(tag: String, message: String, details: String? = null) {
        addLog(LogLevel.NETWORK, tag, message, details)
    }
    
    fun alert(tag: String, message: String, details: String? = null) {
        addLog(LogLevel.ALERT, tag, message, details)
    }
    
    fun command(tag: String, message: String, details: String? = null) {
        addLog(LogLevel.COMMAND, tag, message, details)
    }
    
    fun response(tag: String, message: String, details: String? = null) {
        addLog(LogLevel.RESPONSE, tag, message, details)
    }
    
    // ============================================
    // MÉTODOS ESPECÍFICOS PARA SUPABASE
    // ============================================
    
    fun supabaseRequest(endpoint: String, method: String, body: String? = null) {
        command("SUPABASE", "$method $endpoint", body)
    }
    
    fun supabaseResponse(endpoint: String, code: Int, response: String? = null) {
        if (code in 200..299) {
            response("SUPABASE", "✓ $endpoint -> $code", response?.take(200))
        } else {
            error("SUPABASE", "✗ $endpoint -> $code", response)
        }
    }
    
    fun supabaseError(endpoint: String, error: String) {
        error("SUPABASE", "✗ $endpoint FAILED", error)
    }
    
    // ============================================
    // MÉTODOS ESPECÍFICOS PARA ALERTAS
    // ============================================
    
    fun alertDetected(deviceId: String, temp: Float, message: String) {
        alert("ALERT", "🚨 ALERTA DETECTADA: $deviceId", "Temp: $temp°C - $message")
    }
    
    fun alertSilenced(deviceId: String, source: String) {
        success("ALERT", "🔕 ALERTA SILENCIADA: $deviceId", "Origen: $source")
    }
    
    fun alertResolved(deviceId: String) {
        success("ALERT", "✅ ALERTA RESUELTA: $deviceId", null)
    }
    
    fun silenceCommandSent(deviceId: String) {
        command("SILENCE", "📤 Enviando comando SILENCE", "Device: $deviceId")
    }
    
    fun silenceCommandConfirmed(deviceId: String) {
        success("SILENCE", "✓ Comando SILENCE confirmado", "Device: $deviceId")
    }
    
    fun silenceCommandFailed(deviceId: String, error: String) {
        error("SILENCE", "✗ Comando SILENCE falló", "Device: $deviceId - $error")
    }
    
    // ============================================
    // MÉTODOS ESPECÍFICOS PARA POLLING
    // ============================================
    
    fun pollingStarted() {
        info("POLLING", "▶️ Polling iniciado", "Intervalo: 5 segundos")
    }
    
    fun pollingResult(devicesCount: Int, hasAlert: Boolean) {
        val status = if (hasAlert) "⚠️ CON ALERTA" else "✅ Sin alertas"
        info("POLLING", "📊 $devicesCount dispositivo(s) - $status", null)
    }
    
    fun pollingError(error: String) {
        error("POLLING", "✗ Error en polling", error)
    }
}
